src/app/beer-list/beer-list/beer-list-datasource.ts

```diff
@@ -0,0 +1,56 @@
+import { DataSource } from '@angular/cdk/collections';
+import { MatSort } from '@angular/material';
+import { map, startWith, switchMap } from 'rxjs/operators';
+import { Observable, merge } from 'rxjs';
+import { BeerStoreService } from '../../beer-store.service';
+import { Beer } from '../../app.types';
+import { LoadingService } from '../../loading.service';
+
+export class BeerListDataSource extends DataSource<Beer> {
+
+    private unsetLoading: () => void;
+
+    constructor (private sort: MatSort, private beerStoreService: BeerStoreService, private loadingService: LoadingService) {
+        super();
+    }
+
+    connect (): Observable<Beer[]> {
+        const displayDataChanges = [
+            this.sort.sortChange
+        ];
+        return merge(...displayDataChanges).pipe(
+            startWith(null),
+            switchMap(() => {
+                this.unsetLoading = this.loadingService.setLoading();
+                return this.beerStoreService.getBeers();
```

Comment:
does this get called everytime the sorting changes? what if I already have all the data? why does this component load again then sort the data?

Reply:
Yep, everytime you sort it reloads the data, which is not needed for sorting, since it loads all the data.